app/controllers/whats_app/webhook_controller.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+module WhatsApp
+  class WebhookController < ApplicationController
+    skip_before_action :require_login, :verify_authenticity_token
+    UNSUCCESSFUL_DELIVERY = %w[undelivered failed].freeze
+
+    def message
+      adapter = WhatsAppAdapter::Inbound.new
+
+      adapter.on(WhatsAppAdapter::UNKNOWN_CONTRIBUTOR) do |whats_app_phone_number|
+        exception = WhatsAppAdapter::UnknownContributorError.new(whats_app_phone_number: whats_app_phone_number)
+        ErrorNotifier.report(exception)
+      end
+
+      adapter.on(WhatsAppAdapter::RESPONDING_TO_TEMPLATE_MESSAGE) do |contributor|
+        contributor.update!(whats_app_message_template_responded_at: Time.current, whats_app_template_message_sent_at: nil)
+        message = contributor.received_messages.first
+        WhatsAppAdapter::Outbound.send!(message)
+      end
+
+      adapter.on(WhatsAppAdapter::UNKNOWN_CONTENT) do |contributor|
+        WhatsAppAdapter::Outbound.send_unknown_content_message!(contributor)
+      end
+
+      whats_app_webhook_params = webhook_params.to_h.transform_keys(&:underscore)
+      adapter.consume(whats_app_webhook_params) { |message| message.contributor.reply(adapter) }
+    end
+
+    def errors
+      return unless webhook_params['Level'] == 'ERROR' && JSON.parse(params['Payload'])['error_code'].to_i == 21_617
+
+      payload = JSON.parse(webhook_params['Payload'])
+      more_info = payload['more_info']
+      parameters = payload['webhook']['request']['parameters']
+      error_text = more_info['Msg'].gsub(Setting.twilio_account_sid, '<ACCOUNT_SID>')
+      exception = WhatsAppAdapter::MaximumMessageLengthExceededError.new(text: error_text,
+                                                                         server_phone_number: parameters['channelToAddress'])
+      ErrorNotifier.report(exception, context: { error_sid: webhook_params['Sid'], message_sid: parameters['messageSid'] })
+    end
+
+    def status
+      return unless status_params['MessageStatus'].in?(UNSUCCESSFUL_DELIVERY)
+
+      exception = WhatsAppAdapter::MessageDeliveryUnsuccessfulError.new(status: status_params['MessageStatus'],
+                                                                        whats_app_phone_number: status_params['To'].split('whatsapp:').last)
+      ErrorNotifier.report(exception, context: { message_sid: status_params['MessageSid'] })
+    end
+
+    private
+
+    def webhook_params
+      params.permit(:AccountSid, :ApiVersion, :Body, :From, :Level, :Latitude, :Longitude, :MessageSid, :NumMedia, :NumSegments,
+                    :ParentAccountSid, :Payload, :PayloadType, :ProfileName, :ReferralNumMedia, :Sid, :SmsMessageSid, :SmsSid,
+                    :SmsStatus, :Timestamp, :To, :WaId, *media_params)
```

Comment:
Why is `media_params` in the list of strong params? Isn't that a method?

Reply:
I can't remember where I saw this trick, but it's a way of permitting multiple strong params returned by the `media_params` method as an array. The idea was that if we received three images, they would come as:

```
MediaUrl0 MediaContentType0
MediaUrl1 MediaContentType1
MediaUrl2 MediaContentType2
```
and we would want to permit all these params. In this scenario, we would also not know how many images we would be receiving per webhook.

Since then, I've re-read the documentation and based on my expercience, we only ever receive one "media" per webhook, so I've simplified the logic in 3424a1f3